typed-protocols/test/ChannelTests.hs

```diff
@@ -0,0 +1,135 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module ChannelTests (tests) where
+
+import Control.Monad.Class.MonadAsync
+import Control.Monad.Class.MonadTime
+import Control.Monad.IOSim
+
+import Network.TypedProtocol.Channel
+
+import Data.Time.Clock (diffTimeToPicoseconds, picosecondsToDiffTime)
+import System.Random
+
+import Test.QuickCheck
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.QuickCheck (testProperty)
+
+
+tests :: TestTree
+tests =
+  testGroup "Channel"
+  [ testProperty "createConnectedDelayChannels" prop_createConnectedDelayChannels
+  ]
+
+
+prop_createConnectedDelayChannels
+  :: Positive Int
+  -> (Positive Int, Positive Int, Positive Int)
+  -> Int
+  -> [Int] -> Property
+prop_createConnectedDelayChannels
+    (Positive maxsize) (Positive g_ms, Positive s_ms, Positive v_ms)
+    seed xs =
+
+      expectedDelayChannelTimes maxsize (g, s, v) prng xs
+  === actualDelayChannelTimes   maxsize (g, s, v) prng xs
+  where
+    prng = mkStdGen seed
+
+    g = millisecondsToDiffTime g_ms
+    s = millisecondsToDiffTime s_ms
+    v = millisecondsToDiffTime v_ms
+
+    millisecondsToDiffTime :: Int -> DiffTime
+    millisecondsToDiffTime = (/1e3) . fromIntegral
+
+actualDelayChannelTimes
+  :: RandomGen g
+  => Int
+  -> (DiffTime, DiffTime, DiffTime)
+  -> g
+  -> [b]
+  -> [((VTime, b), (VTime, b))]
+actualDelayChannelTimes maxsize gsv prng xs0 =
+    runSimOrThrow $ do
+      (chanA, chanB) <- createConnectedBufferedChannels
+                          (const 1) maxsize gsv prng
+
+      sa <- async (sender chanA xs0)
+      ra <- async (receiver chanB [] (length xs0))
+      uncurry zip <$> waitBoth sa ra
+  where
+    sender chan xs =
+      sequence
+        [ do send chan x
+             now <- getMonotonicTime
+             return (now,x)
+        | x <- xs ]
+
+    receiver _    xs 0 = return (reverse xs)
+    receiver chan xs n = do
+      Just x <- recv chan
+      now <- getMonotonicTime
+      receiver chan ((now,x):xs) (n-1)
+
+
+expectedDelayChannelTimes
+  :: RandomGen g
+  => Int
+  -> (DiffTime, DiffTime, DiffTime)
+  -> g
+  -> [b]
+  -> [((VTime, b), (VTime, b))]
+expectedDelayChannelTimes maxsize (g, s, v) prng0 xs0 =
```

Comment:
It's actually surprising to me that one can dequeue only when the channel is full to correctly model channels with delay. I think that's possible because the time (the `now` argument of `go`) is decoupled and moves on as we `dequeue`.

Maybe you have a simple model how this works, it would be good to put it in haddock.

Reply:
My "ah ha" moment along these lines was realizing that the scheduled "arrival times" of the elements in the buffer might not be monotonic -- so they're actually the "earliest possible arrival time" per message, not necessarily the actual arrival time. Hence ``arrive = max maxarrive ((g + vsample) `addTime` depart)``.

That contention/overlap between the queue's semantics and the scheduled arrival times' values was a bit surprising. Would it be worthwhile to maintain the `max` while generating the arrival times? Or perhaps just a comment/renaming some variables would suffice. (Possibly include whichever in the implementation too, so that debugging the queue's contents might be less confusing?)

More directly to Marcin's comment: `to` is called once per element, as expected. The queue maintained here is _not_ analogous to the original queue of elements. A renaming might help: perhaps `now -> senderTime` and `maxarrive` to `receiverTime`? Or `senderClock` and `receiverClock`, etc.